### PR TITLE
feat: Extensibility for Appsync API

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of changes
- added amplify override types
- amplify override command
- Migration workflow
- Added/Updated unity tests
- Added e2e test for override

#### Checklist

- [X] PR description included
- [X] `yarn test` passes
- [X] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts
Comment (jhockett) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/service-walkthroughs/appSync-walkthrough.ts:159]: Why did printer changes get reverted? Rebasing issue maybe?

File: packages/amplify-graphql-function-transformer/src/__tests__/overrides/build/override.js
Comment (jhockett) [inline packages/amplify-graphql-function-transformer/src/__tests__/overrides/build/override.js:1]: Should this file be in git?

File: packages/amplify-graphql-http-transformer/src/__tests__/amplify-graphql-http-transformer.test.ts
Comment (jhockett) [inline packages/amplify-graphql-http-transformer/src/__tests__/amplify-graphql-http-transformer.test.ts:30]: Debugging leftover?

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/src/transformation/transform.ts:292]: remove commented console.logs

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/src/transformation/transform.ts:341]: same here

File: packages/amplify-graphql-transformer-core/src/transformation/transform.ts
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/src/transformation/transform.ts:368]: remove if not needed

File: packages/amplify-graphql-transformer-core/src/types/utils.ts
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/src/types/utils.ts:93]: nit: Could use `$TSObject` instead of `any`. Here and in the functions below

File: packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts
Comment (jhockett) [inline packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts:226]: nit: remove if no longer needed

File: packages/amplify-graphql-transformer-core/package.json
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/package.json:41]: Use the same version as the rest of the CLI - `"~1.124.0"`.

**Review by jhockett — CHANGES_REQUESTED** [review]

File: packages/amplify-category-api/src/__tests__/provider-utils/awscloudformation/cfn-api-artifact-handler.test.ts
Comment (akshbhu) [reply packages/amplify-category-api/src/__tests__/provider-utils/awscloudformation/cfn-api-artifact-handler.test.ts:250]: this test isnt needed since resolver config is not part of transformer.conf.json

File: packages/amplify-category-api/src/provider-utils/awscloudformation/cfn-api-artifact-handler.ts
Comment (akshbhu) [reply packages/amplify-category-api/src/provider-utils/awscloudformation/cfn-api-artifact-handler.ts]: done

File: packages/amplify-graphql-function-transformer/src/__tests__/overrides/build/override.js
Comment (akshbhu) [reply packages/amplify-graphql-function-transformer/src/__tests__/overrides/build/override.js:1]: this is used in unit test for transformers

File: packages/amplify-graphql-transformer-core/src/types/utils.ts
Comment (akshbhu) [reply packages/amplify-graphql-transformer-core/src/types/utils.ts:93]: I have used any since this package is not dependent on amplify-cli-core

File: packages/amplify-category-api/src/__tests__/provider-utils/awscloudformation/cfn-api-artifact-handler.test.ts
Comment (jhockett) [reply packages/amplify-category-api/src/__tests__/provider-utils/awscloudformation/cfn-api-artifact-handler.test.ts:250]: In that case, we should remove it, but I won't block the merge for that

**Review by jhockett — APPROVED** [review]

File: packages/amplify-graphql-transformer-core/package.json
Comment (jhockett) [inline packages/amplify-graphql-transformer-core/package.json:41]: @akshbhu this has the wrong version prefix. It must be `~`. See the changes in yarn.lock for why

Comment (akshbhu) [reply packages/amplify-graphql-transformer-core/package.json]: done

**Review by sachscode — APPROVED** [review]
LGTM

File: packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts:332]: nit: name this variable `paths` since it can be files or dirs

File: packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/graphql-transformer/utils.ts:331]: is the purpose of this function to empty everything except the tsconfig.resource.json file? If so, I'd rename this function to `emptyBuildDirRetainTsconfig` But also, why does the tsconfig need to be retained? Seems like all files in a build directory should be ephemeral

**Review by edwardfoyle — CHANGES_REQUESTED** [review]
One high-level thing: I'd do a pass through this PR and anywhere that "migration" is mentioned add some wording to make it clear it is an "extensibility migration" or "overrides migration". This is to disambiguate it from the transformer 1 => 2 migration that also exists now.

There are also several places where we have a tagged release version referenced in a package.json file that needs to be fixed before merging

**Review by edwardfoyle — COMMENTED** [review]
...and I just realized this PR is going into a feature branch not master so I guess depending on the tagged release versions is okay for now. Still needs to be fixed before merging to master, but I won't block this PR on that.

**Comment by akshbhu** [comment]
Closing this in favour of https://github.com/aws-amplify/amplify-cli/pull/8916

**Pull request closed without merging**
